### PR TITLE
Totp change

### PR DESCRIPTION
**Pull request by SleeplessOne1917**
## Description

This is an update so the UI works with the 2FA changes on the server side.

## Screenshots

This is what shows when enabling 2FA:
[enable_totp_demo.webm](https://github.com/LemmyNet/lemmy-ui/assets/28871516/9a8479fb-a9b7-4425-8801-55718f109354)

This video shows using the new modal on the login screen:
[totp_login_demo.webm](https://github.com/LemmyNet/lemmy-ui/assets/28871516/39b71c2b-47a2-4c47-a960-43438c0a1d79)

This video shows how the modal works when entering 1 character at a time and hitting backspace, as well as what happens when an incorrect token is entered:
[totp_play_around.webm](https://github.com/LemmyNet/lemmy-ui/assets/28871516/dacdaf62-693c-452c-883a-9f2d39d5eec8)

## TODO
Add translations

### CODE REVIEW

File: src/assets/css/main.css [hunk 1]
```diff
@@ -448,3 +448,7 @@ br.big {
 .skip-link:focus {
   top: 0;
 }
+
+.totp-link {
+  width: fit-content;
```
Comment (dessalines) [inline src/assets/css/main.css:453]: I couldn't find any bootstrap class for this, oh well.

File: src/shared/components/common/totp-modal.tsx [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+        aria-labelledby="#totpModalTitle"
+        data-bs-backdrop="static"
+        ref={this.modalDivRef}
+      >
+        <div className="modal-dialog modal-fullscreen-sm-down">
+          <div className="modal-content">
+            <header className="modal-header">
+              <h3 className="modal-title" id="totpModalTitle">
+                {type === "generate"
+                  ? "Enable 2 Factor Authentication"
```
Comment (dessalines) [inline src/shared/components/common/totp-modal.tsx:165]: You already noted this, but ya these need i18n.

File: src/shared/components/common/totp-modal.tsx [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+      this.setState({
+        qrCode: URL.createObjectURL(
+          new Blob([(await getSVG(this.props.secretUrl!)).buffer], {
+            type: "image/svg+xml",
+          }),
+        ),
+      });
+    }
+  }
+}
```
Comment (dessalines) [inline src/shared/components/common/totp-modal.tsx:266]: It looks really great, just wish we didn't have to rely on bootstrap modals, and so much custom dom manipulation.

File: src/shared/components/person/settings.tsx [hunk 4]
```diff
@@ -1001,57 +1042,90 @@ export class Settings extends Component<any, SettingsState> {
...
+      const siteRes = await HttpService.client.getSite();
+      UserService.Instance.myUserInfo!.local_user_view.local_user.totp_2fa_enabled =
+        enabled;
+
+      if (siteRes.state === "success") {
+        this.setState({ siteRes: siteRes.data });
+      }
+
+      toast(
+        `Successfully ${
```
Comment (dessalines) [inline src/shared/components/person/settings.tsx:1109]: Also i18n.

**Review by dessalines — COMMENTED** [review]
I tested this locally also, and it works great, thx a ton!

Just add the i18n stuff and we'll get it merged.

File: src/shared/components/common/totp-modal.tsx (hunk 3 above)
Comment (SleeplessOne1917) [reply src/shared/components/common/totp-modal.tsx:268]: I did a bit of refactoring to make handling the modal a bit more react-like.

**Review by dessalines — APPROVED** [review]
GJ. I grepped and couldn't find any other of those states.

**Comment by dessalines** [comment]
I'll fix that and merge.

**Pull request merged**